src/hashtable.py

```diff
@@ -114,23 +125,36 @@ def retrieve(self, key):
         else:
             return None
 
+    def _transfer_storage(self):
+        old_storage = self.storage
+        self.storage = [None] * self.capacity
+
+        for item in old_storage:
+            while item:
+                    # using insert method takes care of hashing
+                self.insert(item.key, item.value)
+                self.keys -= 1  # remove double counting from transfer insert
+                item = item.next
+
     def resize(self):
         '''
         Doubles the capacity of the hash table and
         rehash all key/value pairs.
 
         Fill this in.
         '''
-        self.capacity *= 2
-        old_storage = self.storage
-        self.storage = [None] * self.capacity
+        load_factor = self.keys / self.capacity
+        resized = self.capacity > self.original_capacity
 
-        for item in old_storage:
-            while item:
-                # using insert method takes care of hashing
-                self.insert(item.key, item.value)
-                item = item.next
-        return
+        if load_factor > 0.7:
+            self.capacity *= 2
+            self._transfer_storage()
+
+        elif load_factor < 0.2 and resized:
+            print("resized", load_factor, len(self.storage))
+            self.capacity //= 2
+            self._transfer_storage()
+            print("resized after", len(self.storage))
 
```

Comment:
Nice job handling both the min and max load factor as well as abstracting the `transfer_storage()` logic to a helper function.